### PR TITLE
[#261] Add permission request handler

### PR DESCRIPTION
**Pull request by ryan-conway**
https://github.com/nimblehq/android-templates/issues/261

## What happened 👀

- Added `Peko` dependency to template
- Updated XML and Compose samples
  - Added `requestPermissions` `StateFlow` and `onPermissionResult()` to `HomeViewModel` and `HomeComposeViewModel`
  - Added `requestPermissions()` method to `HomeFragment` and `HomeComposeFragment`

## Insight 📝

Since the template currently targets `SDK 30` instead of `SDK 31`, we must use an older version of Peko (Current stable is `2.2.0`, we are using `2.1.2`). This version makes use of Coroutines and LiveData, and the library must be called from an Activity or Fragment (We may call `Peko.requestPermissionsAsync()` if we have a Context in our ViewModel, but this could lead to leaks)

If we wish to use Flow with permission requests, we should use Peko `3.X` (latest is `3.0.0-ALPHA-01`). This would require updating target SDK version to `31`, as well as updating Kotlin and Coroutines version to `v1.6.21`.

The logic and code is the same for both XML and Compose, will update the `HomeComposeFragment` and `HomeComposeViewModel` once code for XML is finalized 👍🏻 

## Proof Of Work 📹

[untitled.webm](https://user-images.githubusercontent.com/53168251/196604648-ab9314c4-a5e8-4adc-abf8-81255c42e225.webm)

### CODE REVIEW

File: template/app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -3,6 +3,8 @@
     package="co.nimblehq.template">
 
     <uses-permission android:name="android.permission.INTERNET" />
+    <uses-permission android:name="android.permission.ACCESS_FINE_LOCATION" />
+    <uses-permission android:name="android.permission.CAMERA" />
```
Comment (Tuubz) [inline template/app/src/main/AndroidManifest.xml:7]: This feels like sample code, would you agree? 💭

File: template/app/src/main/java/co/nimblehq/template/ui/screens/xml/HomeViewModel.kt [hunk 2]
```diff
@@ -31,6 +39,27 @@ class HomeViewModel @Inject constructor(
...
+        )
+    }
+
+    fun onPermissionResult(permissionResult: PermissionResult) {
+        when (permissionResult) {
+            is Granted -> Timber.d("${permissionResult.grantedPermissions} granted")
+            is Denied.JustDenied -> Timber.d("${permissionResult.deniedPermissions} denied")
+            is Denied.NeedsRationale -> Timber.d("${permissionResult.deniedPermissions} needs rationale")
+            is Denied.DeniedPermanently -> Timber.d("${permissionResult.deniedPermissions} denied for good")
+            is Cancelled -> Timber.d("request cancelled")
```
Comment (Tuubz) [inline template/app/src/main/java/co/nimblehq/template/ui/screens/xml/HomeViewModel.kt:62]: Same here, I'm wondering if we should only keep the dependency in `template` and move the rest into the `samples` 

What do you think? 🙏

**Review by Tuubz — COMMENTED** [review]
Rest LGTM 💪

File: template/app/src/main/AndroidManifest.xml (hunk 1 above)
Comment (ryan-conway) [reply template/app/src/main/AndroidManifest.xml:7]: Good point 👍🏻 Shall we also update `HomeViewModel.requestPermissions()` to emit an empty array instead? Or remove that method entirely? 🤔

Comment (Tuubz) [reply template/app/src/main/AndroidManifest.xml:7]: I'm fine to keep `HomeViewModel.requestPermissions()` as it is (for the `samples`) 💪

Comment (ryan-conway) [reply template/app/src/main/AndroidManifest.xml:7]: If thats the case then running the project after generating could lead to a crash because those permissions weren't declared in the manifest 🤔

Comment (Tuubz) [reply template/app/src/main/AndroidManifest.xml:7]: Can we do this?

- `template`: Only define the dependency, nothing else
- `samples`: Define dependency, declare permissions, integrate home-fragment & home-view model

Please let me know if I got the wrong idea here 🙇

Comment (ryan-conway) [reply template/app/src/main/AndroidManifest.xml:7]: That could definitely work 👀 Will apply it like that 😄

File: template/app/src/main/AndroidManifest.xml [hunk 3]
```diff
@@ -14,7 +14,7 @@
         android:supportsRtl="true"
         android:theme="@style/AppTheme">
 
-        <activity android:name=".ui.screens.MainActivity">
+        <activity android:name=".ui.screens.MainActivity" android:exported="true">
```
Comment (Tuubz) [inline template/app/src/main/AndroidManifest.xml:17]: Is this change necessary for this pull request? 🙏

**Review by Tuubz — APPROVED** [review]
Rest LGTM 🎸

File: sample-xml/app/src/main/java/co/nimblehq/sample/xml/ui/screens/home/HomeViewModel.kt [hunk 4]
```diff
@@ -32,10 +40,31 @@ class HomeViewModel @Inject constructor(
...
+
+    fun onPermissionResult(permissionResult: PermissionResult) {
+        when (permissionResult) {
+            is Granted -> Timber.d("${permissionResult.grantedPermissions} granted")
+            is Denied.JustDenied -> Timber.d("${permissionResult.deniedPermissions} denied")
+            is Denied.NeedsRationale -> Timber.d("${permissionResult.deniedPermissions} needs rationale")
+            is Denied.DeniedPermanently -> Timber.d("${permissionResult.deniedPermissions} denied for good")
+            is Cancelled -> Timber.d("request cancelled")
+        }
+    }
```
Comment (luongvo) [inline sample-xml/app/src/main/java/co/nimblehq/sample/xml/ui/screens/home/HomeViewModel.kt:69]: I personally think permission request is a UI action, so the logic to request permission should be in the View layer 💭 What do you think? What's the benefit of keeping it in VM layer?

Comment (luongvo) [reply sample-xml/app/src/main/java/co/nimblehq/sample/xml/ui/screens/home/HomeViewModel.kt:69]: `onPermissionResult` could be in VM layer 💭

**Review by lydiasama — APPROVED** [review]

File: sample-xml/app/src/main/java/co/nimblehq/sample/xml/ui/screens/home/HomeViewModel.kt [hunk 5]
```diff
@@ -20,6 +24,10 @@ class HomeViewModel @Inject constructor(
     val uiModels: StateFlow<List<UiModel>>
         get() = _uiModels
 
+    private val _requestPermissions = MutableStateFlow<Array<out String>>(emptyArray())
+    val requestPermissions: StateFlow<Array<out String>>
+        get() = _requestPermissions
```
[the remark this answers is not in the document]
Comment (ryan-conway) [reply sample-xml/app/src/main/java/co/nimblehq/sample/xml/ui/screens/home/HomeViewModel.kt:29]: Shouldn't the view be kept as simple as possible with all logic in the ViewModel? 🤔 
Suppose we have a button that needs to check `LOCATION` permission, so the view's code would look something like
```kotlin
button.setOnClickListener {
    if (hasPermission(LOCATION)) {
         doSomething()
    } else {
        requestPermission(LOCATION)
    }
}
```
instead of just 
```kotlin
button.setOnClickListener { viewModel.doSomething() }
```
Please correct me if I've misunderstood something 😅

[the remark this answers is not in the document]
Comment (Tuubz) [reply sample-xml/app/src/main/java/co/nimblehq/sample/xml/ui/screens/home/HomeViewModel.kt:29]: What if we do this in `HomeFragment.kt?` So no code in `HomeViewmodel.kt` as @luongvo suggested

If the user already has permission, it will go to `Granted`, then we can still call `viewModel.doSomething()` if needed.

```
override fun setupView() {
      button.setOnClickListener { 
         requestPermissions(
               Manifest.permission.ACCESS_FINE_LOCATION,
               Manifest.permission.CAMERA
          ) 
     }
}

private fun requestPermissions(vararg permissions: String) {
    lifecycleScope.launch {
        val result = requestPermissionsAsync(*permissions)
        handlePermissionResult(result)
    }
}

private fun handlePermissionResult(permissionResult: PermissionResult) {
    when (permissionResult) {
        is Granted -> Timber.d("${permissionResult.grantedPermissions} granted")
        is Denied.JustDenied -> Timber.d("${permissionResult.deniedPermissions} denied")
        is Denied.NeedsRationale -> Timber.d("${permissionResult.deniedPermissions} needs rationale")
        is Denied.DeniedPermanently -> Timber.d("${permissionResult.deniedPermissions} denied for good")
        is Cancelled -> Timber.d("request cancelled")
    }
}

```

File: sample-compose/app/src/main/java/co/nimblehq/sample/compose/ui/screens/home/HomeComposeViewModel.kt [hunk 6]
```diff
@@ -33,10 +41,31 @@ class HomeComposeViewModel @Inject constructor(
...
+        )
+    }
+
+    fun onPermissionResult(permissionResult: PermissionResult) {
+        when (permissionResult) {
+            is Granted -> Timber.d("${permissionResult.grantedPermissions} granted")
+            is Denied.JustDenied -> Timber.d("${permissionResult.deniedPermissions} denied")
+            is Denied.NeedsRationale -> Timber.d("${permissionResult.deniedPermissions} needs rationale")
+            is Denied.DeniedPermanently -> Timber.d("${permissionResult.deniedPermissions} denied for good")
+            is Cancelled -> Timber.d("request cancelled")
```
Comment (Tuubz) [inline sample-compose/app/src/main/java/co/nimblehq/sample/compose/ui/screens/home/HomeComposeViewModel.kt:68]: @ryan-conway Any idea how to trigger `Cancelled` & `JustDenied`? I can trigger the rest correctly though

Comment (ryan-conway) [reply sample-compose/app/src/main/java/co/nimblehq/sample/compose/ui/screens/home/HomeComposeViewModel.kt:68]: According the the [readme for v2.1.2](https://github.com/deva666/Peko/tree/release/2.1.2)
>PermissionResult.Denied.JustDenied -> subclass of PermissionResult.Denied, returned when previous two cases are not the cause, for example if you forget to register the Permission in AndroidManifest
PermissionResult.Cancelled -> returned when Android System cancels the request, ie returned

Comment (Tuubz) [reply sample-compose/app/src/main/java/co/nimblehq/sample/compose/ui/screens/home/HomeComposeViewModel.kt:68]: Nice, thank you 🙏 

Also, when I tap outside the permission dialog, it results in `DeniedPermanently`.

Not what I was expecting... What do you think? 💭

Comment (ryan-conway) [reply sample-compose/app/src/main/java/co/nimblehq/sample/compose/ui/screens/home/HomeComposeViewModel.kt:68]: Looks like this is an [existing issue](https://github.com/deva666/Peko/issues/21) with the library 👀 
Not sure if this has been fixed in the latest version, but even if it was we can't use it anyway until we update our dependencies 🥲 
Given that, shall we still continue to use the library? Seems like a pretty big issue to have considering this is meant to be a good template to use 😅

Comment (Tuubz) [reply sample-compose/app/src/main/java/co/nimblehq/sample/compose/ui/screens/home/HomeComposeViewModel.kt:68]: Ughhhh, that sucks 😭 

I'm afraid we'll have to put this on hold for now and wait until we can upgrade to `SDK 31` (Probably in December)

Hopefully, the issue has been fixed by then.

If you agree, you can put this pull request as a draft 🪦

Comment (ryan-conway) [reply sample-compose/app/src/main/java/co/nimblehq/sample/compose/ui/screens/home/HomeComposeViewModel.kt:68]: Agreed, will draft it for now 🥲

**Comment by Tuubz** [comment]
@luongvo Any last word on this pull request? 🤩

File: sample-compose/app/src/main/java/co/nimblehq/sample/compose/ui/screens/home/HomeComposeFragment.kt [hunk 7]
```diff
@@ -25,8 +35,31 @@ class HomeComposeFragment : BaseComposeFragment() {
             )
         }
 
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+        requestPermissions(
+            Manifest.permission.ACCESS_FINE_LOCATION,
+            Manifest.permission.CAMERA
```
Comment (chornerman) [inline sample-compose/app/src/main/java/co/nimblehq/sample/compose/ui/screens/home/HomeComposeFragment.kt:42]: (Optional) Use static import
```suggestion
            ACCESS_FINE_LOCATION,
            CAMERA
```

**Review by chornerman — COMMENTED** [review]
The rest LGTM

**Comment by Tuubz** [comment]
@chornerman I've updated the pull request, since @ryan-conway will be off today 🚀

**Review by chornerman — APPROVED** [review]
LGTM 🚀

**Review by luongvo — APPROVED** [review]
LGTM
